internal/workspace/queries.go

```diff
@@ -0,0 +1,144 @@
+/*
+MIT License
+
+Copyright (c) 2025 jupyter-ai-contrib
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+*/
+
+package workspace
+
+import (
+	"context"
+	"fmt"
+
+	workspacev1alpha1 "github.com/jupyter-ai-contrib/jupyter-k8s/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// LabelWorkspaceTemplate is the label key for tracking which template a workspace uses
+// LabelWorkspaceTemplateNamespace is the label key for tracking which namespace the template reference uses
+// These are defined here to avoid import cycles with the controller package
+const (
+	LabelWorkspaceTemplate          = "workspace.jupyter.org/template"
+	LabelWorkspaceTemplateNamespace = "workspace.jupyter.org/template-namespace"
+)
+
+// ListByTemplate returns all active workspaces using the specified template
+// Supports pagination for large-scale deployments
+// Filters out workspaces being deleted (DeletionTimestamp set)
+// Validates templateRef matches label to guard against drift
+// If templateNamespace is empty, it acts as a wildcard (backwards compatible)
+func ListByTemplate(ctx context.Context, k8sClient client.Client, templateName string, templateNamespace string, continueToken string, limit int64) ([]workspacev1alpha1.Workspace, string, error) {
+	logger := logf.FromContext(ctx)
+
+	workspaceList := &workspacev1alpha1.WorkspaceList{}
+
+	// Build label selector - namespace is optional for backwards compatibility
+	labels := map[string]string{
+		LabelWorkspaceTemplate: templateName,
+	}
+	if templateNamespace != "" {
+		labels[LabelWorkspaceTemplateNamespace] = templateNamespace
+	}
+
+	listOptions := []client.ListOption{
+		client.MatchingLabels(labels),
+	}
+
+	// Add pagination options if specified
+	if limit > 0 {
+		listOptions = append(listOptions, client.Limit(limit))
+	}
+	if continueToken != "" {
+		listOptions = append(listOptions, client.Continue(continueToken))
+	}
+
+	if err := k8sClient.List(ctx, workspaceList, listOptions...); err != nil {
+		return nil, "", fmt.Errorf("failed to list workspaces by template label: %w", err)
+	}
+
+	// Filter out workspaces being deleted and verify template reference
+	// This follows Kubernetes controller best practice: resources with deletionTimestamp
+	// are considered "gone" for dependency checking purposes
+	activeWorkspaces := []workspacev1alpha1.Workspace{}
+	for _, ws := range workspaceList.Items {
```

Comment:
ah I see, we can't just pass `Limit=1` then...Anyway we could filter on `DeletionTimestamp==0` ?

If no, let's discuss whether we need pagination here.

Also, could you rename `ListNonDeletedWorkspaceByTemplate()` or smth that conveys that logic please?